modules/javafx.graphics/src/main/java/javafx/scene/Node.java

```diff
@@ -10459,4 +10438,38 @@ void releaseAccessible() {
         }
     }
 
+    /**
+     * Requests focus traversal from this {@code Node} in the specified direction.
+     * A successful traversal results in the newly focused {@code Node} visibly indicating its focused state.
```

Comment:
What's the rationale behind always setting the `focusVisible` bit when doing programmatic focus traversal? Visible focus is always cleared when calling the existing `requestFocus()` method, since it is specified to _only_ be set when focus is acquired by keyboard navigation.

Maybe we should have the next node visibly indicate focus if (and only if) the current node visibly indicates focus. This would be consistent with the [Selectors 4 spec](https://www.w3.org/TR/selectors-4/#the-focus-visible-pseudo):
> If the previously-focused element indicated focus, and a script causes focus to move elsewhere, the newly focused element should indicate focus.

> 
> Conversely, if the previously-focused element did not indicate focus, and a script causes focus to move elsewhere, the newly focused element should also not indicate focus.

Reply:
The rationale is that this PR does not change the way the current focus subsystem works, only provides the public API.

Reply:
So you're proposing to change the specification of `focusVisible`. That needs to be discussed first.

Reply:
Because `focusVisible` is specified to _only_ be set as a result of keyboard navigation, not as a result of programmatic navigation.

`Node.focusVisible` specifies:
```
    /**
     * Indicates whether this {@code Node} should visibly indicate focus.
     * This flag is set when the node acquires input focus via keyboard navigation,
     * and it is cleared when the node loses focus or when {@link #requestFocus()}
     * is called.
     *
     * @defaultValue false
     * @since 19
     */
```

The [JBS issue](https://bugs.openjdk.org/browse/JDK-8268225) says:
> For JavaFX, "Node.focusVisible" would be set in addition to "Node.focused" if a scene graph node gained focus as a consequence of keyboard input. In all other cases, "Node.focusVisible" would not be set on a node.

Reply:
To clarify, this PR changes nothing in the way FX behaves in relation to focus.  Nothing.  I only provided a public API for use in custom controls. 

So whatever discrepancy between FX and w3.org spec you mention (not sure why, this isn't an FX spec) applies equally to the existing implementation (if not, please let me know - you found a bug).

However, perhaps, you pointing out that the docs may be improved - and you are doing so in a very difficult manner.  Or maybe you are pointing out the fact that we should bring back the traverse(Node, TraversalDirection, boolean focusVisible) call?

Reply:
I just quoted you the JavaFX specification, not W3C. It very clearly says that the `focusVisible` flag is only set when a node acquires focus as a result of keyboard navigation. You are proposing to extend the API to change focus programmatically, and in the process also set `focusVisible`, thus violating that specification.

Reply:
Maybe we should say that `Node.requestFocusTraversal()` should only be called by custom components/skins as a response to key press?  I can't see a way to enforce that particular rule.

Or perhaps `Node.focusVisible` doc should be updated to include the reference to `Node.requestFocusTraversal()` ?

What's your opinion?

Reply:
Let's consider a few scenarios:

1. A node has acquired focus via keyboard navigation. Its `focusVisible` bit is set. When I press "tab", I expect the next node to also have visible focus. This is the status quo.
2. Same as 1, but instead of pressing "tab", I press "enter", which has a custom handler that changes focus with the new `requestFocusTraversal` API. I think is is quite natural to assume that the visible bit will also carry over to the new node.
3. A node acquires focus with mouse navigation. Its `focusVisible` bit is not set. When pressing "tab", I expect the next node to have the visible bit.
4. Same as 3, but instead of pressing "tab", I press "enter". The custom handler changes focus programmatically, and the new node does not have the visible bit.

It seems we have the same problem that we originally had with `requestFocus()`, namely that we can't distinguish whether the API was called in response to a key press, or in response to any other event (mouse, programmatically, etc.).

Back when `focusVisible` was added, we decided to always clear the visible bit, because this seemed more correct than to always set it. Like you say, there's no way to tell why someone called `requestFocusTraversal`, and thus we can never clearly know whether to set or to clear the visible bit.

Since we can't know, maybe the right course of action is to allow the caller to specify whether to clear or set the visible bit, because the caller should know that. The API could then be something like: `requestFocusTraveral(TraversalDirection, boolean visible)`.

Reply:
I think the javadoc for `Node.focusVisible` (and its sibling `Node.focusWithin`) is rather insufficient.  The JBS ticket is not a normative document, perhaps this should be clarified in a follow-up ticket (by moving some of the verbiage from JDK-8268225 ?)

I am not against adding a boolean for `visible` (or `focusVisible` ?) argument.

@mstr2 what would be a good description for this parameter?

Reply:
What do you think of the following:
```
* @param visible Specifies whether the {@link #focusVisibleProperty() focusVisible} flag will be
*                set on the node that receives focus. Callers must specify {@code true} if this
*                method is called as a result of keyboard navigation, or if the current node
*                visibly indicates focus; in all other cases, callers must specify {@code false}.
```

Reply:
I like that, thank you!

Reply:
also updated the JEP:
https://github.com/andy-goryachev-oracle/Test/blob/main/doc/FocusTraversal/FocusTraversal-v3.md

Reply:
Somewhat tangential question: should we start placing JEPs in `/doc-files/jeps` ?

The use of private repos is ok at the initial stages of discussion, but at some point we are risking the bit rot: private repos can be removed but the mailing list posts remain...

Reply:
> I like that, it clearly explains what to do. thank you!

I do, too. When I read the discussion up to the point Michael proposed this, my thought was that the only way to solve it is with a parameter like this.

> Somewhat tangential question: should we start placing JEPs in /doc-files/jeps ?

There was a thread a while ago about this. Let's restart that thread rather than discussing it in any particular PR.

Reply:
If this flag is suddenly so important, then why doesn't `requestFocus` have it?

I feel that we've found an internal inconsistency, and that the solution is to then drop this on the user to ensure correctness (the wording "callers must specify" definitely alludes to this).  It's like asking users to pass along some important flag, and if they get it wrong the system breaks.

I think the request focus call should take the most sensible value for focusVisible on programmatic changes, and that FX should not be using this API itself for mouse focus actions.  Now we're just exposing focusVisible in a round-about way (as a boolean flag of some other call) to make it user controllable -- then why not just make it writable?  If I don't like the calculated default (or perhaps always `true`), then I can call `Node#setFocusVisible(false)` after the request focus call.

Reply:
Good question, @hjohn .  Why indeed, @mstr2  ?

Reply:
> If this flag is suddenly so important, then why doesn't `requestFocus` have it?

That's easy: because we didn't deem it necessary. Back when we added the flag, only internal code would be able to set it, and that has worked out fine. Focus traversal using keyboard navigation didn't use `requestFocus()`.

Now we have a public API, and we face the same question again. There is a clear advantage of setting the focusVisible flag as a parameter of `requestFocusTraversal`: we ensure consistency between `focused` and `focusVisible`, such that a node cannot be focusVisible if not focused. Having focusVisible be a writable property would allow applications to have a node visibly indicate focus, or even multiple independent nodes indicating visible focus simultaneously, while yet another node is actually focused. That's not good.

Reply:
How about this for `focusVisible`:
```
    /**
     * Indicates whether this {@code Node} should visibly indicate focus.
     * <p>
     * This flag is set when the node is {@link #focusedProperty() focused} via keyboard navigation,
     * or when the input focus was programmatically moved from another node to this node with
     * {@link #requestFocusTraversal} while the other node visibly indicated focus.
     * <p>
     * The {@code focusVisible} flag can be used by applications that do not want a permanent focus
     * indicator, and opt for a focus indicator that is only visible when it is most helpful to the
     * user. This is not the case when the user clicks on a control (because they would know which
     * control they clicked on), but only when a keyboard-initiated focus traversal moves the input
     * focus to another control.
     */
```

As `requestFocusTraversal` is concerned, I think we can simply this a bit. We don't need users to check whether `focusVisible` is set on the current node; the implementation can do this internally. Here's a revised parameter doc:
```
* @param keyNavigation {@code true} if this method is called as a result of keyboard navigation;
*                      {@code false} otherwise.
```

The implementation would then set `focusVisible` to true if either the `keyNavigation` argument is true, or if the current node's `focusVisible` flag is true.

Reply:
This is better than the previous, since it only asks the caller of this method, typically the Skin of a custom control, to provide the one missing piece of information -- whether or not the method is in response to keyboard navigation -- and not ask the caller to compute the final answer. That's good. The only question I have is: can we do even better? Is there enough information for the implementation of `requestFocusTraversal` to know whether it was called from  keyboard event? If not, then this updated proposal seems reasonable.

Reply:
I think it could be possible, but I'm not sure if it would be pretty.  Scene is the one dispatching these events, and since FX has a single threaded model, Scene could have a flag that indicates it is currently dispatching a key event.  If `requestFocusTraversal` is called, it can detect if this was in response to a key event higher up the call stack by checking the flag on Scene.

I do think there are other mechanisms in FX that pass along information from shallower callstacks to deeper callstacks in some form or another.  `ExpressionHelper` for example is aware of nested calls and takes action accordingly, and I'm almost certain there are more in places were FX is dealing with peers and syncing the scene graph.

Reply:
What happens when a different type of event is created by a `KeyEvent` handler?  How does the downstream code differentiate between the secondary events coming from a key event handler vs. from one originated normally?

> Is there enough information for the implementation of requestFocusTraversal to know whether it was called from keyboard event?

The (original) intent of this API is to enable the custom components trigger the focus traversal as a response to key events only, the problem is that we can't prevent the application code from misusing the API and calling it for other reason.  (There is no other reason for the application code or custom component to call this method with the `visible=false`, other than misuse)

I would rather specify that this method must be called as a response to KeyEvent, and drop the argument.

Reply:
That seems reasonable, given that it is the expected use case. I wouldn't say "_must_ be called", but rather that it "is _expected_ to be called" in response to a KeyEvent, and then document the behavior accordingly -- namely, the node that receives focus will be `focusVisible`.

@mstr2 What do you think?

Reply:
> What happens when a different type of event is created by a `KeyEvent` handler? How does the downstream code differentiate between the secondary events coming from a key event handler vs. from one originated normally?

This wouldn't matter IMHO, if you fire another event, that just deepens the call stack (the secondary event is fired immediately, not after the key event finishes); the original key event will still be the one that initiated it -- ie. user pressed a key, and focus changed as a result -- no matter how that exactly happened it is then FX's decision to turn on focus visible highlighting.  The whole focus visible highlighting is a feature provided by FX to give it a more native look and feel, but it may be platform dependent which is why I think the user should probably not have a say in the matter. 

I already suggested to make the property writable, but if the user can't be trusted to use it correctly, then they can't be trusted to set such a boolean correctly either.

Reply:
Is there really no conceivable scenario in which this method could legitimately be called _not_ in response to a key event, or is it simply that we're not thinking of such a scenario right now? I'd be bothered by an API that was created with a too-narrow focus in mind, and as a result doesn't allow me to do what I need.

Reply:
Could you think of a scenario when we want a component focused but not indicate its focused state?

Reply:
Yes, that's actually the most common case (focused, but not focusVisible) in contemporary user interfaces, see my post below.